RFC: Client Controlled Nullability Operators Implementation

twof opened the pull request:
Implementation of https://github.com/graphql/graphql-spec/issues/867 along with the suggested `?` syntax. 

[Here's a doc with a list of some of the important test cases and the justifications for their behavior.](https://docs.google.com/document/d/1relgXUdhJrlaSR70lrITQKmcIHXjIVyWxDCKRuzMEvs/edit#heading=h.gm7m1frqgtc4)

This PR includes work done by @xuewei8910, @aprilrd, @magicmark, and @lizjakubowski.

twof commented on package.json:
```diff
@@ -63,6 +63,7 @@
     "eslint-plugin-istanbul": "0.1.2",
     "eslint-plugin-node": "11.1.0",
     "eslint-plugin-tsdoc": "0.2.14",
+    "esm": "^3.2.25",
```
I think this can be removed

twof commented on package-lock.json:
```diff
@@ -28,6 +28,7 @@
         "eslint-plugin-istanbul": "0.1.2",
```
We can revert changes to this file

twof commented:
> [<img alt="CLA Missing ID" width="328" height="28" src="https://camo.githubusercontent.com/10030abc9a0bd659c8175562a2f1a9f7786dcd921f5a5f454e6268ec7953fe17/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f636c612d70726f6a6563742d6c6f676f2d70726f642f636c612d6d697373696e672d69642e737667">](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/12261526/38307428/3281/#/?version=2)
> 
> * ❌ The commit ([43e9fd1](https://github.com/graphql/graphql-js/commit/43e9fd1334db7ea0a10d020938b4cc8b61cbda7a) ,c0e1e7b63d0b95ad00e5b5089d6f7cc1aafe6ca7 ,d63f290dde75c0c80352461947d8f09855430e22 ,8d78724e8569db4f173229981a61325e28d2d5e3 ,08f9319f07fadcc5abbace5b839ccf86f735ef41 ,043f1071dbe04d63fdb31896a7b9e683ade575ba ,102586d45f7f1cee51a9bcbf4595eaf5c18ad4c0 ,301a4ab4fa01395cf17b4c74e84f73124bd9f7ab ,2d7f496ae681e2dee7381eb04c76dc5ca4a44800) is missing the User's ID, preventing the EasyCLA check. [Consult GitHub Help](https://help.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-linked-to-the-wrong-user) to resolve.For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).

All of these commits are mine, but they weren't attributed to any github account for one reason or another.

saihaj commented:
@twof try rebasing and overwriting author on commits https://stackoverflow.com/a/54363151/11321732

twof commented:
@saihaj Thanks! That worked.

IvanGoncharov commented:
@twof I enabled CI, and some tests are failing.

twof commented on src/validation/rules/RequiredStatusOnFieldMatchesDefinitionRule.ts:
```diff
@@ -0,0 +1,54 @@
+import type { ValidationContext } from '../ValidationContext';
+import { modifiedOutputType } from '../..';
+
+/**
+ * Fields on correct type
+ *
+ * A GraphQL document is only valid if all fields selected are defined by the
+ * parent type, or are an allowed meta field such as __typename.
+ *
+ * See https://spec.graphql.org/draft/#sec-Field-Selections
+ */
```
cleanup. Comment is cargo culted

twof commented on src/utilities/__tests__/TypeInfo-test.ts:
```diff
@@ -12,6 +12,7 @@ import { getNamedType, isCompositeType } from '../../type/definition';
 
 import { buildSchema } from '../buildASTSchema';
 import { TypeInfo, visitWithTypeInfo } from '../TypeInfo';
+import { ComplexRequiredStatus } from '../../language/ast';
```
I might need a test in this file somewhere

twof commented on src/validation/rules/OverlappingFieldsCanBeMergedRule.ts:
```diff
@@ -590,17 +592,27 @@ function findConflict(
-      [node2],
-    ];
+  if (type1 && type2) {
+    // Errors will have already been handled by RequiredStatusOnFieldMatchesDefinitionRule
+    //  so there's no need to do anything here in the event that modifiedOutputType throws
+    //  an error.
+    try {
+      var modifiedType1 = modifiedOutputType(type1, node1.required);
+      var modifiedType2 = modifiedOutputType(type2, node2.required);
+
+      if (doTypesConflict(modifiedType1, modifiedType2)) {
+        return [
+          [
+            responseName,
+            `they return conflicting types "${inspect(
+              modifiedType1,
+            )}" and "${inspect(modifiedType2)}"`,
+          ],
+          [node1],
+          [node2],
+        ];
+      }
+    } catch {}
```
Is it okay not handling errors here if I expect the errors these functions throw to be handled in a different rule? Is there any reason to think one rule will be run and the other won't?

twof commented on src/validation/rules/RequiredStatusOnFieldMatchesDefinitionRule.ts:
```diff
@@ -0,0 +1,61 @@
+ * A GraphQL document is only valid if all fields selected are defined by the
+ * parent type, or are an allowed meta field such as __typename.
+ *
+ * See https://spec.graphql.org/draft/#sec-Field-Selections
```
Update this link once I've added this rule to the spec

twof commented on src/language/printer.ts:
```diff
@@ -55,9 +56,47 @@ const printDocASTReducer: ASTReducer<string> = {
       const prefix = wrap('', alias, ': ') + name;
-      let argsLine = prefix + wrap('(', join(args, ', '), ')');
+      const requiredDesignator = buildRequiredDesignator(
+        required as unknown as ComplexRequiredStatus,
```
Typescript thinks that `required` is a `string`, but it's actually a `ComplexRequiredStatus`. I'm not sure why that's happening, so this is how I'm dealing with it. Let me know if there's something else I can do.

twof commented:
Hello again! I've implemented the list syntax described here: https://github.com/graphql/graphql-spec/pull/895#issuecomment-963909931

I'd like a code review on this before I add the list syntax to the spec, in case the implementation ends up being dramatically different after I apply any feedback I get. It's a large PR, so if there's any way I can make reviewing this easier, I'd be happy to do so.

IvanGoncharov commented on src/index.ts:
```diff
@@ -236,8 +236,6 @@ export {
   ParseOptions,
   SourceLocation,
-  TokenKindEnum,
-  KindEnum,
```
this is strange.
Problems with rebase?

twof replied:
I think my editor was telling me these were deprecated, but I can try bringing them back.

IvanGoncharov commented on src/language/__tests__/lexer-test.ts:
```diff
@@ -168,36 +168,36 @@ describe('Lexer', () => {
     let caughtError;
     try {
-      lexOne(['', '', '    ?', ''].join('\n'));
+      lexOne(['', '', ' ~', ''].join('\n'));
```
please extract changes in this file as separate PR, happy to merge them as is.
This PR is quite big so let's try to extract parts that don't require spec changes.
It will review simpler for everyone from WG.

Also please add test for `?` similar to what we already have for `!`

twof replied:
I think I've extracted everything I can. Unfortunately it's not a whole lot, but let me know if I've missed anything. I've also added `?` tests to `lexer-test.ts`.

https://github.com/graphql/graphql-js/pull/3281#issuecomment-975831116

twof replied:
Let me know if there are any notes I can write up that might help you and others review this PR. I fully understand that it's quite large!

twof commented:
Extracted changes that don't need to be in this PR:
https://github.com/graphql/graphql-js/pull/3376

IvanGoncharov commented on src/language/ast.ts:
```diff
@@ -360,6 +385,7 @@ export interface FieldNode {
   readonly arguments?: ReadonlyArray<ArgumentNode>;
   readonly directives?: ReadonlyArray<DirectiveNode>;
   readonly selectionSet?: SelectionSetNode;
+  readonly required: ComplexRequiredStatus;
```
This should be a part of AST.
Without it you can't point to error.
For example, I wrote query like that:
```
{ 
   twoDimentionalArray[[[!]!]!]!
}
```
So error should be highlighted as follow:
```
{ 
   twoDimentionalArray[[[!]!]!]!
                        ^^^
}
```
If we go with "support array" root, then we should introduce 3 new AST nodes:
1. for `!`
2. for `?`
3. for `[]`

Also it will allow making property in `FieldNode` optional to represent "normal" state.
```
readonly requireModifiers?: RequireModifierNode;
```

IvanGoncharov replied:
Also it will make `print` implementation simpler and make this feature compatible with other AST based function like `visit`.

IvanGoncharov replied:
Thinking more about it, I'm not sure if how to design AST or if we should have 3 new nodes or one more complex node. 
But 100% sure it should be implemented as recursive AST nodes.
So please feel free to experiment with design and see what will fit our AST design.

IvanGoncharov commented:
> Let me know if there are any notes I can write up that might help you and others review this PR. I fully understand that it's quite large!

@twof Great idea. 
Let's do first round of review where I check "architectural stuff", e.g. AST design.
This round can provoke quite big changes so I suggest writing notes after it finishes.

twof commented:
@IvanGoncharov Sounds like a plan. Thanks!

twof commented:
@IvanGoncharov  I'll take a look at those failing jobs in the morning, but `npm test` is passing on my machine. In the meantime if you'd like to look at the changes I've made and let me know what you think, the architecture is ready for review again.

I'll go over everything and do some cleanup in the morning, so don't worry too much about the details for now. Also let me know if there are any portions you'd like me to write up explanations for.

Summary of changes: 
- `ComplexRequiredStatus` has been broken up into three `Node` interfaces.
- Printing and parsing is now much simpler. You were totally right about that.
- `modifiedOutputType` now also uses a visitor.

Todo:
- It sounds like from the spec feedback that people want the list nullability modifier syntax (`[!]!`) to always match the depth of the type definition for the field it's being applied to. It currently fails if the modifier is too deep (`[[]]`), but it still succeeds if it's too shallow (`!`).
- People haven't landed on behavior for the question mark syntax yet, but that's liable to cause significant changes to this implementation.

twof commented:
@IvanGoncharov Hey Ivan. During the meeting last Thursday you mentioned that you'd be open to releasing this feature behind an experimental flag while it's still in review, and you had some other features that did something similar in the past. Are you able to point me towards an example of a feature that used an experimental flag? I can implement something similar as part of this PR if you're still interested in doing that.

saihaj commented:
Hey @twof we currently publish `defer and stream` as experimental flag you can read more here: https://github.com/graphql/graphql-js#experimental-features. I think for this to happen we first need to get your this branch to the repo. So @IvanGoncharov can help you give rights and set up a branch here which you can push to. Then each time there are changes we just publish a new experimental tag example: https://github.com/graphql/graphql-js/releases/tag/v16.0.0-experimental-stream-defer.5. 

But I think this approach really isn't working well and @IvanGoncharov wanted to merge everything to `main` so we can have unstable `main` and have canary releases. But as per JS-WG it was suppose to happen last Monday or Wednesday (as per a comment I saw) but no updates on that end. But hopefully we can get something this week and have this merged soon cause I want to play around with this feature :) cause I don't want to build myself or start publishing on my fork.

saihaj commented:
> but npm test is passing on my machine

I assume you are on node `v16` cause `v16` tests are the only ones passing.

saihaj commented on src/utilities/applyRequiredStatus.ts:
```diff
@@ -0,0 +1,122 @@
+    ListNullabilityDesignator: {
+      enter() {
+        try {
+          const list = assertListType(getNullableType(typeStack.at(-1)));
```
Might wanna change this cause it is only available in node `v16` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at and we want to support `v12`+

twof replied:
Thanks for pointing that out!

twof commented:
@saihaj Great! Let me know if there's anything I can do to help with the experimental flag setup, or if yall have anything else I can fix up about the PR. It looks like most checks are passing.

IvanGoncharov commented:
@twof I suggest moving in steps.
As the first step let's merge support for `!` and `?` in AST (`parse`, `print`, AST types, `visit`, etc.).
Please split out a smaller PR with just AST changes similar to this one https://github.com/graphql/graphql-js/pull/1141 (note it's more than 4 years old, so code changed a bit since then).
I think one experimental flag for both `!` and `?` would be enough.
Also please make a short summary of the discussion around `!` and `?` syntax (only syntax) and open questions if any and add them in the commit message (or PR message).

That way we will merge noncontroversial stuff first and focus on stuff that needs disscussion.
Also, it will unblock your work with `apollo-ios` and `graphql-codegen`.

twof commented:
@IvanGoncharov Here's the PR with the stuff you asked for: https://github.com/graphql/graphql-js/pull/3418

michaelstaib commented on src/language/ast.ts:
```diff
@@ -360,6 +374,30 @@ export interface FieldNode {
+  | RequiredModifierNode
+  | OptionalModifierNode;
+
+export interface SupportArrayNode {
```
Why is this called SupportArrayNode instead of ListNullabilityNode or at least SupportListNode? In GraphQL we are always talking about lists, not arrays.

twof replied:
The name was based on this comment: https://github.com/graphql/graphql-js/pull/3281#discussion_r754859308

I'm not really a fan of the name I picked either, and I'm down to change it before everything gets merged. `ListNullabilityNode` sounds good to me, so I'll go with that.

michaelstaib commented on src/language/ast.ts:
```diff
@@ -360,6 +374,30 @@ export interface FieldNode {
+  readonly element?: SupportArrayNode;
+}
+
+export interface OptionalModifierNode {
```
Why are the node name and the kind name diverging?

michaelstaib commented on src/language/parser.ts:
```diff
@@ -435,9 +443,77 @@ export class Parser {
       selectionSet: this.peek(TokenKind.BRACE_L)
         ? this.parseSelectionSet()
         : undefined,
+      required: this.parseRequiredStatus(),
```
Shouldn't this come before the selection set and before the directives?

`Field : Alias? Name Arguments? Nullability? Directives? SelectionSet?`


The kitchen sink test never tests a field with selection set:

```
      field3!
       requiredField4: field4!
       field5?
       optionalField6: field6?
       unsetListItemsRequiredList: listField[]!
       requiredListItemsUnsetList: listField[!]
       requiredListItemsRequiredList: listField[!]!
       unsetListItemsOptionalList: listField[]?
       optionalListItemsUnsetList: listField[?]
       optionalListItemsOptionalList: listField[?]?
       multidimensionalList: listField[[[!]!]!]!
```

This must also be tested in its various combinations
```
{
  field? {
    name
  }
}
```

twof replied:
Yeah good point. I'll put up some more tests

michaelstaib replied:
I have fixed this issue in our implementation and have tests now for this.

It must be like the following:

```
return this.node<FieldNode>(start, {
      kind: Kind.FIELD,
      alias,
      name,
      arguments: this.parseArguments(false),
      required: this.parseRequiredStatus(),
      directives: this.parseDirectives(false),
       selectionSet: this.peek(TokenKind.BRACE_L)
         ? this.parseSelectionSet()
         : undefined
```
       
At the moment it would break in your implementation when it is combined with a selection set or if you combine tests with directives.


Here is our implementation of this ... its c# but the order is correct.
https://github.com/ChilliCream/hotchocolate/blob/b1243030b9e20f6c4a22780857e5cd2031d16c68/src/HotChocolate/Language/src/Language.Utf8/Utf8GraphQLParser.Operations.cs#L255

twof replied:
Yep! That's what I went with too

https://github.com/graphql/graphql-js/pull/3418/files#diff-297a41e54c25bbd5c846bb55d7acaaca9f2c9f5fa654b19004da4efbf9a88154R463

twof commented:
Hey @michaelstaib! Thanks for the review. I'm in the process of breaking up this PR into smaller changesets. I've addressed your feedback for the `?` and `!` portions of the proposal in https://github.com/graphql/graphql-js/pull/3418 and your feedback related to the list syntax in https://github.com/twof/graphql-js/pull/2.

@IvanGoncharov @saihaj In order to reduce confusion, would it be helpful to close this PR and designate https://github.com/graphql/graphql-js/pull/3418 as the canonical implementation branch for this RFC?

twof commented on src/execution/execute.ts:
```diff
@@ -489,7 +492,26 @@ function executeField(
   }
 
-  const returnType = fieldDef.type;
+  let returnType: GraphQLOutputType;
```
I don't like this pattern. Reassess.
